pkg/collector/process.go

```diff
@@ -0,0 +1,283 @@
+			if err != nil {
+				return nil, fmt.Errorf("Error in decoding message: %v", err)
+			}
+			elementID[0] = elementID[0] ^ 0x80
+		}
+		field.ElementID = binary.BigEndian.Uint16(elementID)
+		fields = append(fields, &field)
+	}
+	cp.addTemplate(obsDomainID, templateID, fields)
+	return fields, nil
+}
+
+func (cp *collectingProcess) decodeDataRecord(dataBuffer *bytes.Buffer, obsDomainID uint32, templateID uint16) ([]*DataField, error) {
+	template, err := cp.getTemplate(obsDomainID, templateID)
+	if err != nil {
+		return nil, fmt.Errorf("Template %d with obsDomainID %d does not exist", templateID, obsDomainID)
+	}
+	fields := make([]*DataField, 0)
+	for _, templateField := range template {
+		length := int(templateField.ElementLength)
+		field := DataField{}
+		field.Value = bytes.Buffer{}
+		field.Value.Write(dataBuffer.Next(length))
+		field.DataType = cp.getDataType(templateField)
+		fields = append(fields, &field)
+	}
+	return fields, nil
+}
+
+func (cp *collectingProcess) addTemplate(obsDomainID uint32, templateID uint16, fields []*TemplateField) {
+	cp.templatesLock.Lock()
+	if _, exists := cp.templatesMap[obsDomainID]; !exists {
+		cp.templatesMap[obsDomainID] = make(map[uint16][]*TemplateField)
+	}
+	cp.templatesMap[obsDomainID][templateID] = fields
+	cp.templatesLock.Unlock()
+	// template lifetime management
+	if cp.address.Network() == "tcp" {
+		return
+	}
+	if cp.templateTTL == 0 {
+		cp.templateTTL = 1800 * 3 // Default value is 5400s
+	}
+	go func() {
+		ticker := time.NewTicker(time.Duration(cp.templateTTL) * time.Second)
+		defer ticker.Stop()
+		select {
+		case <-ticker.C:
+			klog.Infof("Template with id %d, and obsDomainID %d is expired.", templateID, obsDomainID)
+			cp.deleteTemplate(obsDomainID, templateID)
+			break
+		}
+	}()
+}
+
+func (cp *collectingProcess) getTemplate(obsDomainID uint32, templateID uint16) ([]*TemplateField, error) {
+	cp.templatesLock.RLock()
+	if fields, exists := cp.templatesMap[obsDomainID][templateID]; exists {
+		cp.templatesLock.RUnlock()
+		return fields, nil
+	} else {
+		cp.templatesLock.RUnlock()
+		return fields, fmt.Errorf("Template %d with obsDomainID %d does not exist.", templateID, obsDomainID)
+	}
+}
+
+func (cp *collectingProcess) deleteTemplate(obsDomainID uint32, templateID uint16) {
+	cp.templatesLock.Lock()
+	delete(cp.templatesMap[obsDomainID], templateID)
+	cp.templatesLock.Unlock()
+}
+
+func (cp *collectingProcess) getDataType(templateField *TemplateField) entities.IEDataType {
+	var registry registry.Registry
+	if templateField.EnterpriseID == 0 { // IANA Registry
+		registry = cp.ianaRegistry
+	} else if templateField.EnterpriseID == 55829 { // Antrea Registry
+		registry = cp.antreaRegistry
+	}
+	fieldName, err := registry.GetIENameFromID(templateField.ElementID)
+	if err != nil {
+		klog.Errorf("Information Element with id %d cannot be found.", templateField.ElementID)
+	}
+	ie, _ := registry.GetInfoElement(fieldName)
```

Comment:
May be defining a single map with key as (entID, elementID) and value as information elements better than having a nameID map?
We will have two separate maps: elementName to element, (entID, elementID) to element. This is better than making them different. 
In future, we can turn this in to k8s indexer.